### ISSUE TITLE
Bug: Private Endpoint usage conflicts with ALZ Policy driven management of DNS Zones configuration

### ISSUE THREAD

**Issue by OmnipotentOwl** [Azure/terraform-azurerm-avm-res-keyvault-vault#32]
When attempting to use this module in an environment where DNS Zone configuration is managed by Azure Policy for Private Endpoints this module removes the Policy deployed endpoint private_dns_zone_group configuration which breaks access to the key vault until Azure Policy can restore the configuration. 

Example:
![image](https://github.com/Azure/terraform-azurerm-avm-res-keyvault-vault/assets/1769881/49d66cf7-3507-4f0f-b0de-d11eb8c5d06c)

So far the only way I have found to work around this is to ignore the lifecycle changes on any attribute managed outside of Terraform.

**Comment by OmnipotentOwl** [comment]
Thinking more about this one option to solve this issue could be to create conditional logic in the for_each input for the `azurerm_private_endpoint` whereby if the `private_dns_zone_resource_ids` variable input is provided and not null or not `[]` then the existing logic would apply. However, if the `private_dns_zone_resource_ids` variable is null or `[]` then a different invocation of `azurerm_private_endpoint` would be used where the proper lifecycle ignores is used to handle the proper functionality.

If the above concept were implemented, you would need to stitch the two maps of private endpoints back together to get a unified view of the PEs.

**Comment by matt-FFFFFF** [comment]
Thanks for taking the time to report this issue.

We will discuss internally and get back to you.

/cc @jtracey93 we need to see how this plays with the ALZ policy set. Perhaps a flag to indicate that zone groups are managed externally?

**Comment by jtracey93** [comment]
Yup I think a Boolean flag for zone groups managed by something else is the only way to approach this @matt-FFFFFF

**Comment by matt-FFFFFF** [comment]
@OmnipotentOwl I can repro the issue

- I created a PDNS zone
- I assigned the `Configure Azure Key Vaults to use private DNS zones` policy to the subscription
- I deployed the module using private endpoints, but left the private dns zone resource ids empty
- I waited for DINE
- I ran Terraform plan and observed that the `deployedByPolicy` private dns zone group would be removed

```
  # module.keyvault.azurerm_private_endpoint.this["primary"] will be updated in-place
  ~ resource "azurerm_private_endpoint" "this" {
        id                       = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-ldv2/providers/Microsoft.Network/privateEndpoints/pe-kv-ldv2"
        name                     = "pe-kv-ldv2"
        tags                     = {}
        # (6 unchanged attributes hidden)

      - private_dns_zone_group {
          - id                   = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-ldv2/providers/Microsoft.Network/privateEndpoints/pe-kv-ldv2/privateDnsZoneGroups/deployedByPolicy" -> null
          - name                 = "deployedByPolicy" -> null
          - private_dns_zone_ids = [
              - "/subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/rg-pdns/providers/microsoft.network/privatednszones/privatelink.vaultcore.azure.net",
            ] -> null
        }

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by matt-FFFFFF** [comment]
Options:

1. Use multiple azurerm resources, one with ignore_changes and one without - as per @OmnipotentOwl suggestion
2. Use AzAPI - as the private dns zone group is a child resource, we can decouple the private endpoint and any zone groups. AzAPI would ignore any zone groups it did not know about.

**Comment by OmnipotentOwl** [comment]
Another interesting observation is that depending on how frequent the module executes it can cause the deployment to fail and throw an exception when trying to read the contact properties for the key vault during the refresh operation if the private endpoint hasnt yet been put back by Azure Policy. 
![image](https://github.com/Azure/terraform-azurerm-avm-res-keyvault-vault/assets/1769881/f0663be7-5b5e-4943-bcd4-320c1a70bc6e)

**Comment by matt-FFFFFF** [comment]
Is the above still an issue if you use TF to manage the private dns zone group?

**Comment by kewalaka** [comment]
a related use case that is important for me is the ability to set private dns records on a zone in a "hub" sub - this works at the moment, specifying the resource ID in full, as long as appropriate permissions are in place (which i understand might not be palatable to all).

I couldn't get this use case working with AzAPI, I was getting 'forbidden' errors.  I couldn't see why nor think of any reason why it shouldn't work, but just mentioning it to avoid a potential regression with option 2 here.

**Comment by matt-FFFFFF** [comment]
@OmnipotentOwl do you have an example of the solution that I can see? Would be useful

**Comment by OmnipotentOwl** [comment]
In my case, I had a dedicated Private Endpoint submodule to configure the internal workings of the private endpoint configuration. This gist should give an example of how it worked. For your specific configuration, if you don't want to move to a submodule for the private endpoint configuration, you could use a filtered for_each statement for the two options for managing the private endpoint private DNS group zone.

https://gist.github.com/OmnipotentOwl/f1fb0e368098ac86c8baebed2d5f46c5

**Comment by matt-FFFFFF** [comment]
Thank you!

**Comment by matt-FFFFFF** [comment]
I have raised a PR in the AVM repo to update this - would value your thoughts

https://github.com/Azure/Azure-Verified-Modules/pull/537

**Comment by matt-FFFFFF** [comment]
With the above now closed I can implement the fix here

**Comment by kewalaka** [comment]
@matt-FFFFFF I'm just going through adding this into the template and spotted some issues

the private_endpoint_id needs updating:

```terraform
resource "azurerm_private_endpoint_application_security_group_association" "this"
  for_each = local.private_endpoint_application_security_group_associations

  application_security_group_id = each.value.asg_resource_id
  private_endpoint_id           = azurerm_private_endpoint.this[each.value.pe_key].id
}
```

there is a discrepancy in the outputs in the docs - which looks for "this_managed_dns_zone_group" rather than "this"

**Comment by matt-FFFFFF** [comment]
Thanks @kewalaka 

Will take a look!

**Comment by matt-FFFFFF** [comment]
fixed by #113
